Feature dump-c-type-header flag

mmuesly opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

I want to get the possibility for generating a header describing the internal types of a c module. This allows to access them in a harness without including the c file. Some project decide to define module local types that are only exported by typedefs in a header for the user. This doesn't allow to manipulate fields in the harness, if the harness is compiled without having the c module file in the compile context. Including the c file raises other problems regarding stubbing of functions. 
Therefore, I introduce the `--drop-c-type-header` flag to the `goto-instrument`  binary. This allows to create a header file than can be used to compile the harness. Later, the object files can be linked against the c module object file for CBMC verification.

@tautschnig suggested in a discussion during PR creation to eventually add this flag to the `goto-harness` binary. After careful thinking, I think this is not a good idea, because I don't see how the `dump-*` infrastructure harmonize with the generator driven approach of the goto-harness binary implementation. But I am open for a discussion, where the `--dump-c-type-header` flag should live instead of the `goto-instrument` binary. 


- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

mmuesly committed:
Make dump c behavior configurable

mmuesly committed:
Add extract_type_header.py creating a type header

For some test cases it is required access the module internal data structures from the harness.
This script will read the data types from the goto binary and collect the defines from the c_file.
The result is a type header file that can be included to the harness accessing the module local data structures.
Goto-cc will then link the harness with the original c module for verification.

mmuesly committed:
Add test case for --dump-c-type-header

The global var declarations inside a c module should not be in the header.

mmuesly committed:
Add test case for --dump-c-type-header flag

This is a simple test case showing that the --dump-c-type-header flag for goto-instrument works.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -24,6 +24,36 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "dump_c_class.h"
 #include "goto_program2code.h"
 
+#include <iostream>
```
debugging left-over?

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -282,35 +312,39 @@ void dump_ct::operator()(std::ostream &os)
-    os << global_decl_stream.str() << '\n';
+  if(!global_decl_stream.str().empty() &&
+    dump_c_config.include_global_decl)
+      os << global_decl_stream.str() << '\n';
```
I think clang-format will require some reformatting here, but also: please add braces around the body when either the condition or the body span multiple lines.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -24,6 +24,36 @@ Author: Daniel Kroening, kroening@kroening.com
+  true,
+  true,
+  true,
+  false
```
Please add comments for each member.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -24,6 +24,36 @@ Author: Daniel Kroening, kroening@kroening.com
+  true,
+  true,
+  true,
+  true
```
As above: comments, please.

tautschnig commented on src/goto-instrument/dump_c_class.h:
```diff
@@ -21,6 +21,63 @@ Author: Daniel Kroening, kroening@kroening.com
+struct dump_c_configurationt final
+{
+  /// Include the function declaration in the dump
+  bool include_functions_decl;
```
Maybe `include_function_decls`?

tautschnig commented on src/goto-instrument/dump_c_class.h:
```diff
@@ -21,6 +21,63 @@ Author: Daniel Kroening, kroening@kroening.com
+  bool include_functions_decl;
+
+  /// Include the functions in the dump
+  bool include_functions;
```
Maybe `include_function_bodies`?

tautschnig commented on src/goto-instrument/dump_c_class.h:
```diff
@@ -21,6 +21,63 @@ Author: Daniel Kroening, kroening@kroening.com
+  bool include_functions;
+  
+  /// Include the global declaration in the dump
+  bool include_global_decl;
```
Maybe `include_global_decls`? But also: what exactly is a "global declaration?"

mmuesly replied:
The stream controlled by this flag is called global_decls. It is used to declare struct names borrowed from other header files.

tautschnig commented on scripts/extract_type_header.py:
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
```
Random place: please review the commit message, it has a couple of typos: "required *to* access the *module*", "and *collect*"

tautschnig commented on scripts/extract_type_header.py:
```diff
@@ -0,0 +1,107 @@
+# Dump type header for private types in c modules.
+#
+# Author: Malte Mues <mail.mues@gmail.com>
+# Copyright Amazon, Inc. 2019
```
No need for the Amazon copyright here.

tautschnig commented on scripts/extract_type_header.py:
```diff
@@ -0,0 +1,107 @@
+                       cwd=tmpdir,
+                       shell=False)
+
+        absolut_header = os.path.normpath(os.path.join(tmpdir, header_file))
```
Nit pick: "absolut" isn't a part of the English language, how about simply using `header_path`?

tautschnig commented on scripts/extract_type_header.py:
```diff
@@ -0,0 +1,107 @@
+    with TemporaryDirectory() as tmpdir:
+        module = get_module_name(c_file)
+
+        header_file = "{}_datastructure.h".format(module)
```
Nit pick: `header_file = module + "_datastructure.h"` would do.

tautschnig commented on scripts/extract_type_header.py:
```diff
@@ -0,0 +1,107 @@
+DEFINE_REGEX_HEADER = re.compile(r"\s*#\s*define\s*([\w]+)")
+
+
+def collect_defines(c_file):
```
I believe this function would deserve a comment of its own, describing what exactly it is trying to accomplish and where it might get things wrong (as it makes some guesses).

mmuesly replied:
I added a comment.

tautschnig commented on scripts/extract_type_header.py:
```diff
@@ -0,0 +1,107 @@
+    FILE = sys.argv[2]
+    if len(sys.argv) == 4:
+        TARGET = sys.argv[3]
+    make_header_file(BINARY, FILE, TARGET)
```
Random place: shouldn't at least one of the tests also make use of this script? Otherwise it lives here and no-one known when, where, and how to use it.

mmuesly replied:
I add a test.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -26,33 +27,14 @@ Author: Daniel Kroening, kroening@kroening.com
 
-// clang-format off
+dump_c_configurationt dump_c_configurationt::default_configuration =
+  dump_c_configurationt().disable_include_headers();
```
It seems odd that the "default configuration" is not what the default constructor generates.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -208,6 +192,22 @@ void dump_ct::operator()(std::ostream &os)
         global_decl_stream << "// " << symbol.location << '\n';
 
+        std::string location_file = id2string(symbol.location.get_file());
+        if(location_file[location_file.length() - 1] == 'h')
```
What if `location_file` is empty? And what's the magic behind ending in a `h` - why is that a distinguishing factor?

mmuesly replied:
I am just looking for header files and assume that all of them are ending in a `h`.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -208,6 +192,22 @@ void dump_ct::operator()(std::ostream &os)
+          if(it == header_files.end())
+          {
+            header_files.push_back(location_file);
+            std::vector<char> v;
```
Unused?

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -208,6 +192,22 @@ void dump_ct::operator()(std::ostream &os)
+              << "#include \""
+              << split_string(location_file, '/', true, true).back() << "\"\n";
+          }
+        }
```
How is this related to the commit message?

mmuesly replied:
I will change the commit message.

The idea is to include the transitive header files that are required to make the new header file compile.

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1466,3 +1478,47 @@ void dump_cpp(
   out << goto2cpp;
 }
+
+bool module_local_declaration(const symbolt &symbol, const std::string module){
```
`static`?

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1466,3 +1478,47 @@ void dump_cpp(
+
+bool module_local_declaration(const symbolt &symbol, const std::string module){
+  std::string filename = id2string(symbol.location.get_file());
+  std::string symbol_module = strip_string(id2string(symbol.module));
```
Why is `strip_string` necessary _only_ here?

mmuesly replied:
Where else would you expect it? module is already stripped as it originates from the command line.
Should I re strip it?

tautschnig commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1466,3 +1478,47 @@ void dump_cpp(
+bool module_local_declaration(const symbolt &symbol, const std::string module){
+  std::string filename = id2string(symbol.location.get_file());
+  std::string symbol_module = strip_string(id2string(symbol.module));
+  std::vector<std::string> name_components = split_string(filename, '/', true, true);
```
I believe you want `get_base_name`?

mmuesly replied:
Yes! Thanks, I missed that.

tautschnig commented on src/goto-instrument/dump_c_class.h:
```diff
@@ -74,6 +86,46 @@ struct dump_c_configurationt final
   /// The config used for dump-c-type-header
   static dump_c_configurationt type_header_configuration;
 
+  dump_c_configurationt disable_include_functions_decl(bool value=false){
```
If it's named `disable_...` then I wouldn't expect a parameter that could actually cause the opposite effect.

tautschnig commented on src/util/irep_ids.def:
```diff
@@ -19,6 +19,7 @@ IREP_ID_ONE(x86_extended)
 IREP_ID_TWO(C_source_location, #source_location)
 IREP_ID_TWO(C_end_location, #end_location)
 IREP_ID_TWO(C_is_padding, #is_padding)
+IREP_ID_ONE(C_no_dump)
```
`C_do_not_dump` maybe?

tautschnig reviewed (commented):
I have myself frequently hoped for such a tool, so I much appreciate thus. Just some need to clean things up a bit.

mmuesly committed:
Add --dump-c-type-header to goto-instrument

The --dump-c-type-header flag is going to dump a header file
containing the file local type definitions in a c file.
While it might be desirable to keep a type definition
only in a c module, for some proofs it is required to
include the type definition supporting the harness setup.
If the file local type definition requires to include types defined
in other header files, an include for the other header file
is added as well.

mmuesly committed:
Add test for extract_type_header.py script

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4bdabceeabda4b6fa74b20e79e1180edcf314e8d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115657641

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a46f200055da71ca04e8a70f6e037d230e440fa8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115664621

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: b661ae1cae2323d02b371c18e45f2b68f4c20ea1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115828755
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

mmuesly commented:
@tautschnig thank you for the review! I addressed all of your comments or followed up on them with requests.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 848bb62ee5fe25c8963d4a5adde1b8fb34c432f0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115829189

peterschrammel reviewed (approved):
lgtm

romainbrenguier commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -493,8 +549,16 @@ void dump_ct::convert_compound(
-    typedef_types.find(unresolved);
+  typedef_typest::const_iterator td_entry;
+  for(td_entry = typedef_types.begin(); td_entry != typedef_types.end();
+      td_entry++)
```
use range for

romainbrenguier commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1432,3 +1496,56 @@ void dump_cpp(
+  std::string symbol_module = strip_string(id2string(symbol.module));
+  if(base_name == module && symbol_module == module)
+  {
+    return true;
```
`return (base_name == module && symbol_module == module)`

romainbrenguier commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1432,3 +1496,56 @@ void dump_cpp(
+  symbol_tablet symbol_table = ns.get_symbol_table();
+  for(symbol_tablet::iteratort it = symbol_table.begin();
+      it != symbol_table.end();
+      it++)
```
range for?

mmuesly replied:
I am not aware of telling the ranged for loop how to pick this iterator.
`for(auto it : symbol_table){}` iterates a slightly different data structure.

romainbrenguier replied:
Strange, it should pick up `.begin()` and `.end()`. At least you should use ++it instead of it++.

romainbrenguier commented on src/goto-instrument/dump_c_class.h:
```diff
@@ -21,6 +21,120 @@ Author: Daniel Kroening, kroening@kroening.com
+      include_global_vars(true),
+      include_compounds(true),
+      follow_compounds(true),
+      include_headers(false)
```
prefer in class initializers to member initializers in constructors https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c48-prefer-in-class-initializers-to-member-initializers-in-constructors-for-constant-initializers

romainbrenguier commented on src/goto-instrument/dump_c_class.h:
```diff
@@ -21,6 +21,120 @@ Author: Daniel Kroening, kroening@kroening.com
+      include_global_vars(include_global_vars),
+      include_compounds(include_compounds),
+      follow_compounds(follow_compounds),
+      include_headers(include_headers)
```
I'm not sure this constructor is useful since it seems clearer to initialize a default one and then set the options as needed.

mmuesly commented:
@romainbrenguier thank you for your comments. Wherever I was able to do so, I followed your suggestion.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 004cd60d5140443047cb8a72d8fa6d6662848526).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116683609
